log4j-api/src/main/java/org/apache/logging/log4j/message/ReusableParameterizedMessage.java

```diff
@@ -119,30 +119,28 @@ public Message memento() {
         return new ParameterizedMessage(messagePattern, getTrimmedParams());
     }
 
-    private void init(final String messagePattern, final int argCount, final Object[] paramArray) {
+    private void init(final String messagePattern, final int argCount, final Object[] args) {
         this.varargs = null;
         this.messagePattern = messagePattern;
         this.argCount = argCount;
-        final int placeholderCount = count(messagePattern, indices);
-        initThrowable(paramArray, argCount, placeholderCount);
-        this.usedCount = Math.min(placeholderCount, argCount);
+        this.patternAnalysis = analyzePattern(messagePattern, patternAnalysis);
+        this.throwable = determineThrowable(args, argCount, patternAnalysis.placeholderCount);
     }
 
-    private static int count(final String messagePattern, final int[] indices) {
-        try {
-            // try the fast path first
-            return ParameterFormatter.countArgumentPlaceholders2(messagePattern, indices);
-        } catch (final Exception ex) { // fallback if more than int[] length (256) parameter placeholders
-            return ParameterFormatter.countArgumentPlaceholders(messagePattern);
-        }
+    private static MessagePatternAnalysis analyzePattern(final String pattern, final MessagePatternAnalysis patternAnalysis) {
+        return (patternAnalysis != null && ParameterFormatter.analyzePattern(pattern, patternAnalysis))
+                ? patternAnalysis
+                : ParameterFormatter.analyzePattern(pattern);
```

Comment:
Nice way to bootstrap the pattern analysis data, but I would make sure that:
 - `placeholderCharIndices` has size 10 at least,
 - if the number of indices exceeds a preconfigured threshold (e.g. 256), the `MessagePatternAnalysis` is dropped by `clear()`.

Reply:
See the changes in a996006f6d1fcccd7bd37c4ec5e308ddaf891fb3, please.